include/sympack/symPACKMatrix2D.hpp

```diff
@@ -865,30 +969,149 @@ namespace symPACK{
+          double epsilon = 1.0;
+          bool iscorrect = true;
+          logfileptr->OFS()<<"BUFFER SIZE: "<<n<<std::endl;
+          T * h_actual = new T[n];
+          CUDA_ERROR_CHECK(cudaMemcpy(h_actual, actual, sizeof(T)*n, cudaMemcpyDeviceToHost));
+          for (int i=0; i<n; i++) {
+            if (fabs(h_correct[i] - h_actual[i]) > epsilon && h_correct[i]!=h_actual[i]) {
+              logfileptr->OFS()<<"Buffers are not equal at index "<<i<<std::endl;
+              logfileptr->OFS()<<"Expected "<<h_correct[i]<<", got "<<h_actual[i]<<std::endl;
+              logfileptr->OFS()<<"Difference: "<<h_correct[i]-h_actual[i]<<std::endl;
+              iscorrect = false;
+            }
+          }
+          if (iscorrect)
+            logfileptr->OFS()<<"Buffers are equal"<<std::endl;
+          delete h_actual;
+        }
+#endif
+	
+        void dump_nnz(std::string prelude) {
+            progressptr->OFS()<<prelude<<std::endl;
+            for (int i=0; i<_nnz; i++) {
+                progressptr->OFS()<<_nzval[i]<<",";
+            }
+            progressptr->OFS()<<std::endl;
+        }
+
+#ifdef CUDA_MODE
+        void terminate_fallback_handler() {
+#if UPCXX_SPEC_VERSION >= 20230300
+            std::string err_msg("Error: GPU device segment of " + 
+                        std::to_string(symPACK::gpu_allocator.segment_size()) + 
+                        " bytes exhausted with " + std::to_string(symPACK::gpu_allocator.segment_used()) +                        " bytes in use. Try increasing `-gpu_mem` (see README.md)");
+#else
+            std::string err_msg("Error: GPU device segment exhausted, try increasing `-gpu_mem` (see README.md)");
+#endif
+            throw std::runtime_error(err_msg);
+        }
+    
+#endif
 
         virtual int factorize( TempUpdateBuffers<T> & tmpBuffers) {
           scope_timer(a,blockCell_t::factorize);
 #if defined(_NO_COMPUTATION_)
           return 0;
 #endif
+
           auto snode_size = std::get<0>(_dims);
           auto diag_nzval = _nzval;
+          try {
+            auto potrf_cpu = [snode_size, diag_nzval](){
+                lapack::Potrf( 'U', snode_size, diag_nzval, snode_size);
+            };
+#ifdef CUDA_MODE
+            if (snode_size*snode_size > symPACK::potrf_limit) {
+                
+            dev_ptr d_diag_nzval = symPACK::gpu_allocator.allocate<T>(snode_size*snode_size);
+            
+            /* Couldn't allocate enough space on GPU, so use CPU */
+            if (d_diag_nzval==nullptr) {
+                switch(symPACK::fallback_type) {
+                    case FallbackType::TERMINATE :
+                        terminate_fallback_handler();
+                        break;
+                    case FallbackType::CPU :
+                        potrf_cpu();
+                        break; 
+                }
+                if (symPACK::gpu_verbose) {
+                    symPACK::cpu_ops["potrf"]+=1;
+                } 
```

Comment:
@Hooninator Thanks for adding these "rough" counters. 

Some suggestions for improvement:
1. Move the CPU counter increments into the lambda above that invokes the operation
2. Move the GPU counter increments into the "delay slot" between the lapack/cublas invocation and the cudaDeviceSynchronize() that finishes the operation
3. Create an inline function to factor, eg:
```
// factored definition in Environment.hpp
template <typename Map>
inline void symPACK::counter(Map &map, std::string const &op) {
  if (symPACK::gpu_verbose) map[op]++;
}
// use examples:
symPACK::counter(symPACK::gpu_ops, "potrf");
symPACK::counter(symPACK::cpu_ops, "gemm");
```

Reply:
Fixed in [4d62638](https://github.com/symPACK/symPACK/pull/12/commits/4d62638fe283eff6a4cea951deb5883d520bb1e6)